File: AesCbc256Pbkdf2HmacStringEncryption/AesCbc256Pbkdf2HmacStringEncryption.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Collections;

public class Program {
	public static void Main() {
		Console.WriteLine("AES CBC 256 String encryption with PBKDF2 derived key and HMAC check");
		
		string plaintext = "The quick brown fox jumps over the lazy dog";
    Console.WriteLine("plaintext: " + plaintext);
		string password = "secret password";

		// encryption
    Console.WriteLine("\n* * * Encryption * * *");
    String ciphertextBase64 = aesCbcPbkdf2HmacEncryptToBase64(password, plaintext);
    Console.WriteLine("ciphertext (Base64): " + ciphertextBase64);
    Console.WriteLine("output is (Base64) salt : (Base64) iv : (Base64) ciphertext : (Base64) hmac");
		
		// decryption
		Console.WriteLine("\n* * * Decryption * * *");
    String ciphertextDecryptionBase64 = ciphertextBase64;
		Console.WriteLine("ciphertext (Base64): " + ciphertextDecryptionBase64);
    Console.WriteLine("input is (Base64) salt : (Base64) iv : (Base64) ciphertext : (Base64) hmac");
    String decryptedtext = aesCbcPbkdf2HmacDecryptFromBase64(password, ciphertextDecryptionBase64);
		Console.WriteLine("plaintext: " + decryptedtext);
	}

  static string aesCbcPbkdf2HmacEncryptToBase64(string password, string data) {
    int PBKDF2_ITERATIONS = 15000;
    byte[] salt = GenerateSalt32Byte();
    byte[] keyAesHmac = new byte[64];
    using (var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2_ITERATIONS, HashAlgorithmName.SHA256))
    {
      keyAesHmac = deriveBytes.GetBytes(64);
    }
		byte[] keyAes = new byte[32];
    byte[] keyHmac = new byte[32];
    Buffer.BlockCopy(keyAesHmac, 0, keyAes, 0, 32);
    Buffer.BlockCopy(keyAesHmac, 32, keyHmac, 0, 32);
    byte[] encrypted;
		byte[] IV = GenerateRandomInitvector();
		using(Aes aesAlg = Aes.Create()) {
			aesAlg.Key = keyAes;
      aesAlg.IV = IV;
			aesAlg.Mode = CipherMode.CBC;
			var encryptor = aesAlg.CreateEncryptor(aesAlg.Key, aesAlg.IV);
			using(var msEncrypt = new MemoryStream()) {
				using(var csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write)) {
					using(var swEncrypt = new StreamWriter(csEncrypt)) {
						swEncrypt.Write(data);
					}
					encrypted = msEncrypt.ToArray();
				}
			}
		}
    string ciphertext = Base64Encoding(salt) + ":" + Base64Encoding(IV) + ":" + Base64Encoding(encrypted);
    HMACSHA256 sha256Hmac = new HMACSHA256(keyHmac);
    byte[] hmac = sha256Hmac.ComputeHash(Encoding.UTF8.GetBytes(ciphertext));
    return ciphertext + ":" + Base64Encoding(hmac);
	}
  
	static string aesCbcPbkdf2HmacDecryptFromBase64(string password, string data) {
    int PBKDF2_ITERATIONS = 15000;
    String[] parts = data.Split(':');
    byte[] salt = Base64Decoding(parts[0]);
    byte[] hmac = Base64Decoding(parts[3]);    
    byte[] keyAesHmac = new byte[64];
    using (var deriveBytes = new Rfc2898DeriveBytes(password, salt, PBKDF2_ITERATIONS, HashAlgorithmName.SHA256))
    {
      keyAesHmac = deriveBytes.GetBytes(64);
    }
    byte[] keyAes = new byte[32];
    byte[] keyHmac = new byte[32];
    Buffer.BlockCopy(keyAesHmac, 0, keyAes, 0, 32);
    Buffer.BlockCopy(keyAesHmac, 32, keyHmac, 0, 32);
    // before we decrypt we have to check the hmac
    String hmacToCheck = parts[0] + ':' + parts[1] + ':' + parts[2];
    HMACSHA256 sha256Hmac = new HMACSHA256(keyHmac);
    byte[] hmacCalculated = sha256Hmac.ComputeHash(Encoding.UTF8.GetBytes(hmacToCheck));
    bool equal = ByteArrayCompare(hmac, hmacCalculated);
    if (!equal) {
      Console.WriteLine("Error: HMAC-check failed, no decryption possible");
      return "";
    }
		string decryptedtext;
		using(Aes aesAlg = Aes.Create()) {
			aesAlg.Key = keyAes;
			byte[] IV = Base64Decoding(parts[1]);
			string encryptedData = parts[2];
			byte[] cipherText = Base64Decoding(encryptedData);
			aesAlg.IV = IV;
			aesAlg.Mode = CipherMode.CBC;
			ICryptoTransform decryptor = aesAlg.CreateDecryptor(aesAlg.Key, aesAlg.IV);
			using(var msDecrypt = new MemoryStream(cipherText)) {
				using(var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Read)) {
					using(var srDecrypt = new StreamReader(csDecrypt)) {
						decryptedtext = srDecrypt.ReadToEnd();
					}
				}
			}
		}
		return decryptedtext;
	}

  static byte[] GenerateSalt32Byte() {
    RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
    byte[] salt = new byte[32];
    rngCsp.GetBytes(salt);
    return salt;
  }

  static byte[] GenerateRandomInitvector() {
    RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
    byte[] iv = new byte[16];
    rngCsp.GetBytes(iv);
    return iv;
  }

  static bool ByteArrayCompare(byte[] a1, byte[] a2) 
  {
    IStructuralEquatable eqa1 = a1;
    return eqa1.Equals(a2, StructuralComparisons.StructuralEqualityComparer);
  }

  static string Base64Encoding(byte[] input) {
    return Convert.ToBase64String(input);
  }

  static byte[] Base64Decoding(String input) {
    return Convert.FromBase64String(input);
  }
}

File: AesCbc256Pbkdf2HmacStringEncryption/AesCbc256Pbkdf2HmacStringEncryption.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:collection/collection.dart';
import "package:pointycastle/export.dart";

main() {
/* add in pubspec.yaml:
dependencies:
  pointycastle: ^3.1.1
 */

  printC('AES CBC 256 String encryption with PBKDF2 derived key and HMAC check');
  final plaintext = 'The quick brown fox jumps over the lazy dog';
  printC('plaintext: ' + plaintext);
  final password = 'secret password';

  // encryption
  printC('\n* * * Encryption * * *');
  String ciphertextHmacBase64 = aesCbcPbkdf2HmacEncryptToBase64(password, plaintext);
  printC('ciphertextHmac (Base64): ' + ciphertextHmacBase64);
  printC('output is (Base64) salt : (Base64) iv : (Base64) ciphertext : (Base64) hmac');

  printC('\n* * * Decryption * * *');
  var ciphertextHmacDecryptionBase64 = ciphertextHmacBase64;
  printC('ciphertextHmac (Base64): ' + ciphertextHmacDecryptionBase64);
  printC('input is (Base64) salt : (Base64) iv : (Base64) ciphertext : (Base64) hmac');
  var decryptedtext = aesCbcPbkdf2HmacDecryptFromBase64(password, ciphertextHmacDecryptionBase64);
  printC('plaintext:  ' + decryptedtext);
}

String aesCbcPbkdf2HmacEncryptToBase64(String password, String plaintext) {
  var plaintextUint8 = createUint8ListFromString(plaintext);
  var passphrase =  createUint8ListFromString(password);
  final PBKDF2_ITERATIONS = 15000;
  final salt = generateSalt32Byte();
  KeyDerivator derivator = new PBKDF2KeyDerivator(new HMac(new SHA256Digest(), 64));
  Pbkdf2Parameters params = new Pbkdf2Parameters(salt, PBKDF2_ITERATIONS, 64); // 2 * 32 for aes + hmac
  derivator.init(params);
  final keyAesHmac = derivator.process(passphrase);
  var keyAes = Uint8List.sublistView(keyAesHmac, 0, 32);
  var keyHmac = Uint8List.sublistView(keyAesHmac, 32, 64);
  final iv = generateRandomIv();
  final CBCBlockCipher cipher = new CBCBlockCipher(new AESFastEngine());
  ParametersWithIV<KeyParameter> cbcParams = new ParametersWithIV<KeyParameter>(new KeyParameter(keyAes), iv);
  PaddedBlockCipherParameters<ParametersWithIV<KeyParameter>, Null> paddingParams = new PaddedBlockCipherParameters<ParametersWithIV<KeyParameter>, Null>(cbcParams, null);
  PaddedBlockCipherImpl paddingCipher = new PaddedBlockCipherImpl(new PKCS7Padding(), cipher);
  paddingCipher.init(true, paddingParams);
  final ciphertext = paddingCipher.process(plaintextUint8);
  final saltBase64 = base64Encoding(salt);
  final ivBase64 = base64Encoding(iv);
  final ciphertextBase64 = base64Encoding(ciphertext);
  // calculate hmac over salt, iv and ciphertext
  final ciphertextWithoutHmac = createUint8ListFromString(saltBase64 + ':' + ivBase64 + ':' + ciphertextBase64);
  final hmac = HMac(SHA256Digest(), 64) // for HMAC SHA-256, block length must be 64
    ..init(KeyParameter(keyHmac));
  final hmacBase64 = base64Encoding(hmac.process(ciphertextWithoutHmac));
  return saltBase64 + ':' + ivBase64 + ':' + ciphertextBase64 + ':' + hmacBase64;
}

String aesCbcPbkdf2HmacDecryptFromBase64(String password, String data) {
  var parts = data.split(':');
  var salt = base64Decoding(parts[0]);
  var iv = base64Decoding(parts[1]);
  var ciphertext = base64Decoding(parts[2]);
  var hmac = base64Decoding(parts[3]);
  var passphrase =  createUint8ListFromString(password);
  final PBKDF2_ITERATIONS = 15000;
  KeyDerivator derivator = new PBKDF2KeyDerivator(new HMac(new SHA256Digest(), 64));
  Pbkdf2Parameters params = new Pbkdf2Parameters(salt, PBKDF2_ITERATIONS, 64); // 2 * 32 for aes + hmac
  derivator.init(params);
  final keyAesHmac = derivator.process(passphrase);
  var keyAes = Uint8List.sublistView(keyAesHmac, 0, 32);
  var keyHmac = Uint8List.sublistView(keyAesHmac, 32, 64);
  // before we decrypt we have to check the hmac
  final ciphertextWithoutHmac = createUint8ListFromString(base64Encoding(salt) + ':' + base64Encoding(iv) + ':' + base64Encoding(ciphertext));
  final hmacCheck = HMac(SHA256Digest(), 64) // for HMAC SHA-256, block length must be 64
    ..init(KeyParameter(keyHmac));
  final hmacCalculated = hmacCheck.process(ciphertextWithoutHmac);
  Function eq = const ListEquality().equals;
  final hmacVerified = eq(hmac, hmacCalculated);
  if (!hmacVerified) {
    printC('Error: HMAC-check failed, no decryption possible');
    return '';
  }
  CBCBlockCipher cipher = new CBCBlockCipher(new AESFastEngine());
  ParametersWithIV<KeyParameter> cbcParams = new ParametersWithIV<KeyParameter>(new KeyParameter(keyAes), iv);
  PaddedBlockCipherParameters<ParametersWithIV<KeyParameter>, Null> paddingParams = new PaddedBlockCipherParameters<ParametersWithIV<KeyParameter>, Null>(cbcParams, null);
  PaddedBlockCipherImpl paddingCipher = new PaddedBlockCipherImpl(new PKCS7Padding(), cipher);
  paddingCipher.init(false, paddingParams);
  return new String.fromCharCodes(paddingCipher.process(ciphertext));
}

Uint8List generateSalt32Byte() {
  final _sGen = Random.secure();
  final _seed =
  Uint8List.fromList(List.generate(32, (n) => _sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(_seed));
  return sec.nextBytes(32);
}

Uint8List generateRandomIv() {
  final _sGen = Random.secure();
  final _seed =
  Uint8List.fromList(List.generate(32, (n) => _sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(_seed));
  return sec.nextBytes(16);
}

Uint8List createUint8ListFromString(String s) {
  var ret = new Uint8List(s.length);
  for (var i = 0; i < s.length; i++) {
    ret[i] = s.codeUnitAt(i);
  }
  return ret;
}

String base64Encoding(Uint8List input) {
  return base64.encode(input);
}

Uint8List base64Decoding(String input) {
  return base64.decode(input);
}

void printC(String newString) {
  // for compatibility reasons to FlutterEmptyConsole
  print(newString);
}
